kythe/go/util/riegeli/writer.go

```diff
@@ -0,0 +1,202 @@
+		b.pos += n
+		if err != nil {
+			return
+		}
+	}
+
+	// For each remaining slice of the chunk of size usableBlockSize, write a
+	// blockHeader and that slice of data.
+	for i, blocksLeft := remainingBlockSize, blockHeaders; blocksLeft > 0; i, blocksLeft = i+usableBlockSize, blocksLeft-1 {
+		// write blockHeader
+		blockStart := b.pos
+		n, err := (&blockHeader{
+			PreviousChunk: uint64(blockStart - chunkStart),
+			NextChunk:     uint64(chunkEnd - blockStart),
+		}).WriteTo(b.w)
+		b.pos += n
+		if err != nil {
+			return b.pos - chunkStart, err
+		}
+
+		l := len(chunk) - i
+		if l > usableBlockSize {
+			// write the maximum size of the chunk possible
+			l = usableBlockSize
+		}
+
+		n, err = b.w.Write(chunk[i : i+l])
+		b.pos += n
+		if err != nil {
+			return b.pos - chunkStart, err
+		}
+	}
+
+	return b.pos - chunkStart, nil
+}
+
+// WriteTo implements the io.WriterTo interface for blockHeaders.
+func (b *blockHeader) WriteTo(w io.Writer) (int, error) {
+	var buf [blockHeaderSize]byte
+	binary.LittleEndian.PutUint64(buf[8:], b.PreviousChunk)
+	binary.LittleEndian.PutUint64(buf[16:], b.NextChunk)
+	binary.LittleEndian.PutUint64(buf[:], hashBytes(buf[8:]))
+	return w.Write(buf[:])
+}
+
+func newRecordChunk(opts *WriterOptions) *recordChunk {
+	return &recordChunk{CompressionType: opts.compressionType()}
+}
+
+// put adds a single record to the Riegeli record chunk.
+func (r *recordChunk) put(rec []byte) error {
+	// TODO(schroederc): compression
+	size := uint64(len(rec))
+
+	var buf [binary.MaxVarintLen64]byte
+	n := binary.PutUvarint(buf[:], size)
+
+	r.CompressedSizes = append(r.CompressedSizes, buf[:n]...)
+	r.CompressedValues = append(r.CompressedValues, rec...)
+
+	r.decodedSize += size
+	r.numRecords++
+	return nil
+}
+
+// encode returns the binary-encoding of the Riegeli record chunk.
+func (r *recordChunk) encode() []byte {
+	// TODO(schroederc): compression
+	buf := bytes.NewBuffer(make([]byte, 0, 1+binary.MaxVarintLen64+len(r.CompressedSizes)+len(r.CompressedValues)))
+
+	buf.WriteByte(byte(r.CompressionType))
+
+	var vBuf [binary.MaxVarintLen64]byte
+	n := binary.PutUvarint(vBuf[:], uint64(len(r.CompressedSizes)))
+	buf.Write(vBuf[:n])
+
+	buf.Write(r.CompressedSizes)
+	buf.Write(r.CompressedValues)
+
+	return buf.Bytes()
+}
+
+// WriteTo implements the io.WriterTo interface for chunkHeaders.
+func (h *chunkHeader) WriteTo(w io.Writer) (int, error) {
+	var buf [chunkHeaderSize]byte
+	binary.LittleEndian.PutUint64(buf[8:16], h.DataSize)
+	copy(buf[16:], h.DataHash[:])
+	buf[24] = byte(h.ChunkType)
+	binary.LittleEndian.PutUint64(buf[25:33], h.NumRecords) // overwrite buf[32] below
+	binary.LittleEndian.PutUint64(buf[32:40], h.DecodedDataSize)
+	hash := hashBytes(buf[8:])
+	binary.LittleEndian.PutUint64(buf[:8], hash)
+	return w.Write(buf[:])
+}
+
+// WriteTo writes the chunk to w, given its starting position within w.
+func (c *chunk) WriteTo(w io.Writer, pos int) (int, error) {
```

Comment:
This name collides semantically with `io.WriterTo`, but the latter doesn't take an offset. How about `WriteToOffset` or something like that?